### DISCUSSION TITLE
How to call a method after all lazy resolves are complete

### DISCUSSION THREAD

**Discussion by trevorturk** [rmosolgo/graphql-ruby#4217] [General]
I ran into a tricky issue today and I thought I should share the solution I came up with, in case it's helpful to others, or perhaps someone can suggest a better way!

I have a weather app with a `Weather` model that fetches data from various upstream API endpoints using `Async::HTTP`. I use `lazy_resolve Async::Task, :wait` to issue the requests in parallel, much like one might do with `Typhoeus::Hydra`. 

When those API requests are made, I have an object that keeps count of the cache hit/miss count, which I then save to a database in a background job to avoid excessive writes on the database. 

The issue I ran into is that after all lazy resolves are complete, I need to call `Weather#save_cache_counts`, but I had trouble finding a workable technique. I couldn't get `GraphQL::Schema::FieldExtension#after_resolve ` to help in my case, but perhaps I was misusing it. 

What I ended up hitting on seems to work just fine, but it looks like a bit of a hack. Basically, I create  a `context` object and pass that into `Schema.execute`, then I stuff my `weather` object in there so I can call methods on it after execution. 

```ruby
# app/graphql/schema.rb
class Schema < GraphQL::Schema
  query Types::Query
  lazy_resolve Async::Task, :wait
end

# app/graphql/types/query.rb
class Types::Query < Types::AsyncObject
  field :weather, Types::Weather do
    argument :api_key, String
  end

  def weather(**args)
    weather = Api::Weather.new(args)
    context[:weather] = weather
    weather
  end
end

# app/controllers/graphql_controller.rb
class GraphqlController < ApplicationController
  def execute
    context = {}

    json = Schema.execute(
      params[:query],
      context: context,
      variables: prepare_variables(params[:variables])
    )

    context[:weather].&save_cache_counts

    render json: json
  end
end
```

So, perhaps this will be helpful to others, but I'd also love to hear other ideas worth trying. Thanks!

**Comment by rmosolgo** [comment]
Thanks for sharing what you found! 👍
